File: game_lobby/anagram_server.cpp
anagram_server

Endpoint AnagramServer;

Struct player_data {
  Endpoint player_helper;
  Number score;
}

AnagramServer {
  Map(from: Text, to: Endpoint) waiting_room;
  Map(from: Text, to: Struct player_data) game;
  List (element: Text) solutions;
  Text anagram;
  TrueFalse game_in_session;
  onCreate() {
    game_in_session = False;
  }

  Public Function set_solution(Text init_anagram, List (element: Text) solution){
    anagram = init_anagram;
    solutions = solution;
  }
  Public Function start_game() {
    for (Text username in game) {
          game[username].score = 0;
	  game[username].player_helper.get_new_message("Game has started.");
    }
    game_in_session = True;
  }
  
  Public Function get_game_status() returns TrueFalse {
    return game_in_session;
  }
 
  Public Function return_anagram() returns Text{
    return anagram;
  }

  Public Function return_solutions() returns List(element: Text) {
    return solutions;
  }

  Public Function get_player_count() returns Number {
    return len(game);
  }

  Public Function get_scores() returns Map(from: Text, to: List(element: Text)) {
    return game;
  }

  Public Function add_score(Text username, Number points) returns Number{
    game[username].score += points;
    return game[username].score;
  }

  Public Function add_player(Text username, Endpoint pt) {
    Text message = username + " is in the game room.";
    Struct player_data player;
    player.score = 0;
    player.player_helper = pt;
    game[username] = player;
    broadcastWaitingMessage(message);
  }

  Public Function add_to_waiting(Text username, Endpoint pt) {
    Text message = username + " has entered the waiting room.";
    broadcastWaitingMessage(message);
    waiting_room[username] = pt;
  }

  Public Function remove_from_waiting(Text username) {
    Text message = username + " has left the waiting room.";
    waiting_room.remove(username);
    broadcastWaitingMessage(message);
  }

  Public Function get_score(Text username) returns Number {
    return game[username].score;
  }

  Public Function remove_user(Text username) {
    game.remove(username);
  }

  Public Function broadcastMessage(Text message) {
    for (Text username in game) {
      game[username].player_helper.get_new_message(message);
    }
  }

  Public Function broadcastWaitingMessage(Text message) {
    for (Text username in waiting_room) {
      waiting_room[username].get_new_message(message);
    }
  }

  Public Function end_game() {
    game_in_session = False;
    broadcastMessage("Game Over.");
    for (Text username in game) {
      Text score = username + ' - ' + toText(game[username].score);
      broadcastMessage(score);
    }
    broadcastMessage("Press ENTER to continue.");
    game = {};
  }
}

File: game_lobby/anagram_server.py
from player_emitted import PlayerHelper
from anagram_server_emitted import AnagramServer
import sys
import os
sys.path.append(os.path.join("../../"))
from waldo.lib import Waldo
import time
import random
HOSTNAME = '127.0.0.1'
PLAYER_NUM = 1
GAMETIME = 60
FILENAME = 'solutions.txt'
SOLUTION_START_INDEX = 1

global solution_set

def print_message (endpoint, message):
    print message

def load_solutions():
    solution_file = open(FILENAME, 'r', 0)
    solutions = solution_file.read().splitlines()
    solution_file.close()
    anagram = ""
    solution = []
    global solution_set
    solution_set = {}
    for line in solutions:
        if line == "":
            solution_set[anagram] = solution
            solution = []
        elif line.startswith('--'):
            anagram = line[2:].upper()
        else:
            solution.append(line.upper())

def set_solutions(anagram_server):
    anagram = random.choice([key for key in solution_set])
    solutions = solution_set[anagram]
    anagram_server.set_solution(anagram, solutions)


if __name__ == '__main__':
    anagram_server = Waldo.no_partner_create(AnagramServer)
    print 'GameServer created!'
    Waldo.tcp_accept(PlayerHelper, 'localhost', 6767, anagram_server)
    load_solutions()
    while True:
        print 'Waiting for players....'
        while anagram_server.get_player_count() <= 0:
            time.sleep(1)
        print 'Player entered.  Game will begin in 20 seconds.'
        set_solutions(anagram_server)
        anagram_server.broadcastWaitingMessage('Game will begin in 20 seconds. Type "/ready" to join.')
        time.sleep(20)
        print 'Game has begun!'
        anagram_server.start_game()
        time.sleep(20)
        anagram_server.end_game()
    
